src/server/stream_family.cc

```diff
@@ -876,6 +911,205 @@ OpResult<vector<GroupInfo>> OpListGroups(const DbContext& db_cntx, string_view k
+    raxStop(&ri_cpel);
+  }
+  raxStop(&ri_consumers);
+  ginfo->consumer_info_vec = std::move(consumer_info_vec);
+}
+
+OpResult<StreamInfo> OpStreams(const DbContext& db_cntx, string_view key, EngineShard* shard,
+                               int full, size_t count) {
+  auto& db_slice = shard->db_slice();
+  OpResult<PrimeIterator> res_it = db_slice.Find(db_cntx, key, OBJ_STREAM);
+  if (!res_it)
+    return res_it.status();
+
+  vector<StreamInfo> result;
+  CompactObj& cobj = (*res_it)->second;
+  stream* s = (stream*)cobj.RObjPtr();
+
+  StreamInfo sinfo;
+  sinfo.length = s->length;
+
+  sinfo.radix_tree_keys = raxSize(s->rax_tree);
+  sinfo.radix_tree_nodes = s->rax_tree->numnodes;
+  sinfo.last_generated_id = s->last_id;
+  sinfo.max_deleted_entry_id = s->max_deleted_entry_id;
+  sinfo.entries_added = s->entries_added;
+  sinfo.recorded_first_entry_id = s->first_id;
+  sinfo.groups = s->cgroups ? raxSize(s->cgroups) : 0;
+  sinfo.entries = GetStreamRecords(s, s->first_id, s->last_id, 0, count);
+
+  if (full) {
+    if (s->cgroups) {
+      GroupInfoVec group_info_vec;
+
+      raxIterator ri_cgroups;
+      raxStart(&ri_cgroups, s->cgroups);
+      raxSeek(&ri_cgroups, "^", NULL, 0);
+      while (raxNext(&ri_cgroups)) {
+        streamCG* cg = (streamCG*)ri_cgroups.data;
+        GroupInfo ginfo;
+        ginfo.name.assign(reinterpret_cast<char*>(ri_cgroups.key), ri_cgroups.key_len);
+        ginfo.last_id = cg->last_id;
+        ginfo.consumer_size = raxSize(cg->consumers);
+        ginfo.pending_size = raxSize(cg->pel);
+        ginfo.entries_read = cg->entries_read;
+        ginfo.lag = streamCGLag(s, cg);
+        ginfo.pel_count = raxSize(cg->pel);
+        GetGroupPEL(s, cg, count, &ginfo);
+        GetConsumers(s, cg, count, &ginfo);
+
+        group_info_vec.push_back(ginfo);
+      }
+      raxStop(&ri_cgroups);
+
+      sinfo.cgroups = group_info_vec;
+    }
+  } else {
+    sinfo.groups = s->cgroups ? raxSize(s->cgroups) : 0;
+    vector<Record> first_entry_vector = GetStreamRecords(s, s->first_id, s->last_id, 0, 1);
+    if (first_entry_vector.size() != 0) {
+      sinfo.first_entry = first_entry_vector.at(0);
+    }
+    vector<Record> last_entry_vector = GetStreamRecords(s, s->first_id, s->last_id, 1, 1);
+    if (last_entry_vector.size() != 0) {
+      sinfo.last_entry = last_entry_vector.at(0);
+    }
+  }
+
+  return sinfo;
+}
+
+OpResult<vector<ConsumerInfo>> OpConsumers(const DbContext& db_cntx, EngineShard* shard,
+                                           string_view stream_name, string_view group_name) {
+  auto& db_slice = shard->db_slice();
+  OpResult<PrimeIterator> res_it = db_slice.Find(db_cntx, stream_name, OBJ_STREAM);
+  if (!res_it)
+    return res_it.status();
+
+  vector<ConsumerInfo> result;
+  CompactObj& cobj = (*res_it)->second;
+  stream* s = (stream*)cobj.RObjPtr();
+
+  streamCG* cg = streamLookupCG(s, sdsnewlen(group_name.data(), group_name.length()));
```

Comment:
there is a memory leak here.`sdsnewlen` allocates an sds string that is never freed

Reply:
Have used the tmp_str1 from shard. Hope that is the correct way to do it across.